File: frontend/src/api/knowledge.ts
// @ts-ignore
/* eslint-disable */
import { request, type RequestOptions } from "../request";

/** Publish Analysis Knowledge POST /api/analysis-runs/${param0}/knowledge-entry */
export async function publishAnalysisKnowledge(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.publishAnalysisKnowledgeParams,
  options?: RequestOptions
) {
  const { identity: param0, ...queryParams } = params;
  return request<API.KnowledgeEntryView>(
    `/api/analysis-runs/${param0}/knowledge-entry`,
    {
      method: "POST",
      params: { ...queryParams },
      ...(options || {}),
    }
  );
}

/** Search Knowledge GET /api/knowledge */
export async function searchKnowledge(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.searchKnowledgeParams,
  options?: RequestOptions
) {
  return request<API.KnowledgePage>("/api/knowledge", {
    method: "GET",
    params: {
      // mode has a default value: exact_substring
      mode: "exact_substring",
      // limit has a default value: 20
      limit: "20",
      ...params,
    },
    ...(options || {}),
  });
}

/** Get Knowledge Entry GET /api/knowledge/${param0} */
export async function getKnowledgeEntry(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getKnowledgeEntryParams,
  options?: RequestOptions
) {
  const { identity: param0, ...queryParams } = params;
  return request<API.KnowledgeEntryView>(`/api/knowledge/${param0}`, {
    method: "GET",
    params: { ...queryParams },
    ...(options || {}),
  });
}

/** Index Knowledge Entry POST /api/knowledge/${param0}/semantic-index */
export async function indexKnowledgeEntry(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.indexKnowledgeEntryParams,
  options?: RequestOptions
) {
  const { identity: param0, ...queryParams } = params;
  return request<API.KnowledgeEntryView>(
    `/api/knowledge/${param0}/semantic-index`,
    {
      method: "POST",
      params: { ...queryParams },
      ...(options || {}),
    }
  );
}

/** Query Knowledge POST /api/knowledge/query */
export async function queryKnowledge(
  body: API.CommentCountQuestion | API.EvidenceQuestion,
  options?: RequestOptions
) {
  return request<API.KnowledgeAnswer>("/api/knowledge/query", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    data: body,
    ...(options || {}),
  });
}

File: backend/src/api/routers/knowledge.py
from typing import Annotated, Literal
from uuid import UUID

from fastapi import APIRouter, Query

from api.dependencies import Authenticated, Knowledge
from api.responses import READ_ERROR_CODES, WRITE_ERROR_CODES, error_responses
from knowledge.schemas import KnowledgeAnswer, KnowledgeEntryView, KnowledgePage, KnowledgeQuestion

router = APIRouter(tags=["knowledge"])


@router.post(
    "/api/v1/analysis-runs/{identity}/knowledge-entry",
    response_model=KnowledgeEntryView,
    status_code=201,
    responses=error_responses(*WRITE_ERROR_CODES, 404, 409),
    operation_id="publishAnalysisKnowledge",
)
def publish_analysis_knowledge(
    identity: UUID,
    service: Knowledge,
    owner: Authenticated,
) -> KnowledgeEntryView:
    return service.publish_analysis(identity)


@router.get(
    "/api/v1/knowledge",
    response_model=KnowledgePage,
    responses=error_responses(*READ_ERROR_CODES, 422),
    operation_id="searchKnowledge",
)
def search_knowledge(
    service: Knowledge,
    owner: Authenticated,
    query: Annotated[str | None, Query(min_length=2, max_length=100)] = None,
    event_id: UUID | None = None,
    mode: Literal["exact_substring", "semantic"] = "exact_substring",
    limit: Annotated[int, Query(ge=1, le=20)] = 20,
) -> KnowledgePage:
    return service.search(query, event_id, limit, mode)


@router.post(
    "/api/v1/knowledge/query",
    response_model=KnowledgeAnswer,
    responses=error_responses(*WRITE_ERROR_CODES, 404, 422),
    operation_id="queryKnowledge",
)
def query_knowledge(
    data: KnowledgeQuestion,
    service: Knowledge,
    owner: Authenticated,
) -> KnowledgeAnswer:
    return service.query(data)


@router.post(
    "/api/v1/knowledge/{identity}/semantic-index",
    response_model=KnowledgeEntryView,
    responses=error_responses(*WRITE_ERROR_CODES, 404, 409),
    operation_id="indexKnowledgeEntry",
)
def index_knowledge_entry(
    identity: UUID,
    service: Knowledge,
    owner: Authenticated,
) -> KnowledgeEntryView:
    return service.index(identity)


@router.get(
    "/api/v1/knowledge/{identity}",
    response_model=KnowledgeEntryView,
    responses=error_responses(*READ_ERROR_CODES, 404),
    operation_id="getKnowledgeEntry",
)
def get_knowledge_entry(
    identity: UUID,
    service: Knowledge,
    owner: Authenticated,
) -> KnowledgeEntryView:
    return service.get(identity)
